### ISSUE TITLE
Detect resolved promises at the end of tests that weren't acted on

### ISSUE THREAD

**Issue by odejesush** [web-platform-tests/wpt#18621]
This is just an idea, but it would be nice if the test framework can produce warnings for tests that contain promises that have resolved after the test has ended. Using the following test as an example:

```js
bluetooth_test(async () => {
  const {device, fake_peripheral} = await getConnectedHealthThermometerDevice();

  const onDisconnected = async () => {
    device.removeEventListener('gattserverdisconnected', onDisconnected);
    // 2. Reconnect.
    await fake_peripheral.setNextGATTConnectionResponse({
      code: HCI_SUCCESS,
    });
    await device.gatt.connect();

    // 3. Disconnect after reconnecting.
    const disconnectPromise = eventPromise(device, 'gattserverdisconnected');
    fake_peripheral.simulateGATTDisconnection();
    await disconnectPromise;
  };
  device.addEventListener('gattserverdisconnected', onDisconnected);

  // 1. Disconnect.
  await fake_peripheral.simulateGATTDisconnection();
}, test_desc);
```

This test doesn't actually test that the Bluetooth device can reconnect and disconnect again because I accidentally forgot to wrap the event listener in a promise, but the test passes successfully. So it would be useful if the test framework can produce a warning that `disconnectPromise` resolved but was not used in the test.

```js
bluetooth_test(async () => {
  const {device, fake_peripheral} = await getConnectedHealthThermometerDevice();

  let reconnectPromise = new Promise(async (resolve) => {
    device.addEventListener('gattserverdisconnected', async () => {
      // 2. Reconnect.
      await fake_peripheral.setNextGATTConnectionResponse({
        code: HCI_SUCCESS,
      });
      await device.gatt.connect();

      // 3. Disconnect after reconnecting.
      const disconnectPromise = eventPromise(device, 'gattserverdisconnected');
      fake_peripheral.simulateGATTDisconnection();
      resolve();
      await disconnectPromise;
    }, {once: true});
  });

  // 1. Disconnect.
  await fake_peripheral.simulateGATTDisconnection();
  await reconnectPromise;
}, test_desc);
```

This is the fixed test that waits until the event listener handler is finished before completing the test.

**Comment by jugglinmike** [comment]
Is it the promise created by the `onDisconnected` function that you have in mind?

Separately, your second example includes the following:

>     resolve();
>     await disconnectPromise;

This code occurs within a DOM event listener, so I don't think the `await` expression will influence execution.

**Comment by odejesush** [comment]
Yes that's the one. I accidentally messed up the "fixed" example. It should end with
```js
  resolve(disconnectPromise);
```
so that the test actually waits until we get a `gattserverdisconnected`. `await` doesn't influence the execution unless I wrap the whole DOM event listener inside of a promise that we `await` at the end.

This does show how easy it is to accidentally end the test before promises that you actually want to see resolved.

**Comment by jugglinmike** [comment]
Yeah, that can be tricky. In general, you can avoid a lot of mistakes by limiting the code which runs in a callback.

```js
bluetooth_test(async () => {
  const {device, fake_peripheral} = await getConnectedHealthThermometerDevice();

  let reconnectPromise = new Promise((resolve) => {
    device.addEventListener('gattserverdisconnected', resolve, {once: true});
  });

  // 1. Disconnect.
  await fake_peripheral.simulateGATTDisconnection();
  await reconnectPromise;

  // 2. Reconnect.
  await fake_peripheral.setNextGATTConnectionResponse({
    code: HCI_SUCCESS,
  });
  await device.gatt.connect();

  // 3. Disconnect after reconnecting.
  const disconnectPromise = eventPromise(device, 'gattserverdisconnected');
  fake_peripheral.simulateGATTDisconnection();
  await disconnectPromise;
}, test_desc);
```

That approach makes the test more sequential (the numbered comments are now self-evident), and it reduces the potential for mistakes.

It also reduces the need for additional explicit error handling. In the prior version, any rejection in the event listener will not be detected by the framework, and the test would fail slowly with a generic timeout instead of quickly with a specific error description.

That said, there's plenty of room for improvement in the harness! Can you say more about how the framework could track the fulfilment of the promise created by the `onDisconnected` function?

**Comment by odejesush** [comment]
Ah, thanks for the example. It does make the test much better. I was going to suggest maybe having an API that will do the same as
```js
let reconnectPromise = new Promise((resolve) => {
  device.addEventListener('gattserverdisconnected', resolve, {once: true});
});
```
but then I saw the [EventWatcher](https://web-platform-tests.org/writing-tests/testharness-api.html#eventwatcher) constructor function which already does something similar. I apologize for that, I should have read through the documentation more carefully. With EventWatcher, the test can be written like this:
```js
async function simulateDisconnection(eventWatcher, fake_peripheral) {
  const disconnectPromise = eventWatcher.wait_for('gattserverdisconnected');
  await fake_peripheral.simulateGATTDisconnection();
  return disconnectPromise;
}
  
bluetooth_test(async (t) => {
  const {device, fake_peripheral} = await getConnectedHealthThermometerDevice();
  const eventWatcher = new EventWatcher(t, device, ['gattserverdisconnected']);

  // 1. Disconnect.
  await simulateDisconnection(eventWatcher, fake_peripheral);

  // 2. Reconnect.
  await fake_peripheral.setNextGATTConnectionResponse({
    code: HCI_SUCCESS,
  });
  await device.gatt.connect();
  
  // 1. Disconnect after reconnecting.
  await simulateDisconnection(eventWatcher, fake_peripheral);
}, test_desc);
```

As for tracking the fulfillment of promises, I'm not really sure how they could be detected. If there was some way to keep track of the promises created, perhaps we could check if the `then()` method was called on them. But after seeing your example above and discovering the `EventWatcher`, I think that these are the best way to write tests like this and prevent this mistake from happening.

**Comment by odejesush** [comment]
Another way that we could perhaps detect unused promises is some sort of static analysis tool.

**Comment by guest271314** [comment]
@odejesush 

> As for tracking the fulfillment of promises, I'm not really sure how they could be detected. If there was some way to keep track of the promises created, perhaps we could check if the then() method was called on them.

If the `Promise` is expected to be fulfilled within a range of time `Promise.race()` can be utilized.

```
let test = new Promise(r => setTimeout(r, 1000, "test"));
let timer = new Promise(r => setTimeout(r, 999, "timer"));
Promise.race([test, timer]).then(value => {
  if (value === "timer") {
    console.log(value); // time fulfilled before test
    return f;
  }
})
.then(console.log) // test fulfilled
.catch(console.error); // handle error
```

**Comment by foolip** [comment]
> Another way that we could perhaps detect unused promises is some sort of static analysis tool.

I've found [eslint-plugin-promise](https://github.com/xjamundx/eslint-plugin-promise) to catch some mistakes with Promises in [Confluence](https://github.com/GoogleChromeLabs/confluence/tree/eslint-plugin-promise).

However, the original problem here was an async event handler, and I suspect that isn't covered by that linter.

Wanting to wait for an event as a promise and wanting to take some async action in response to an event are fairly common patterns in WPT and yet pretty clunky to do. Perhaps improved helpers for this could address most of the problem?
